libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -26,29 +29,81 @@ import kotlin.coroutines.resume
  * This class is responsible for refreshing the current direction route's traffic.
  * This does not support alternative routes.
  */
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
 internal class RouteRefreshController(
     private val routeRefreshOptions: RouteRefreshOptions,
     private val routeRefresh: RouteRefresh,
     private val currentIndicesProvider: CurrentIndicesProvider,
     private val routeDiffProvider: DirectionsRouteDiffProvider = DirectionsRouteDiffProvider(),
-    private val localDateProvider: () -> Date
+    private val localDateProvider: () -> Date,
 ) {
 
+    private var state: RouteRefreshStateResult? = null
+        set(value) {
+            if (field == value) return
+            field = value
+            value?.let { nonNullValue ->
+                observers.forEach {
+                    it.onNewState(nonNullValue)
+                }
+            }
+        }
+
+    private val observers = mutableSetOf<RouteRefreshStatesObserver>()
```

Comment:
I was told you used copy-on-wrote collections so that the observer can unregister themselves in a callback method.

Reply:
We already have UiThread annotations on top of `MapboxNavigation`, so probably we don't need to follow this recommendation at any point

Reply:
That's not the issue. If you remove the observer while iterating over the collection, you get `ConcurrentModificationException`.

Reply:
it's not possible if both operations running on the same thread, is it?

Reply:
I mean sth like that: https://pl.kotl.in/1aGScx9VC

Reply:
I see what you mean 👍 but we don't do something like this in current PR, do you think there's any risk exist?

Reply:
It's not about us doing it, but the external consumers of the callback. We should use the copy-on-write collection here.

Reply:
discussed internally